go/k8s/controllers/user_controller.go

```diff
@@ -0,0 +1,235 @@
+//+kubebuilder:rbac:groups=backend.almond.stanford.edu,resources=users/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the User object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *UserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	backendURL, userID, err := r.getBackendURL(req.NamespacedName.Name, ctx, req)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	userState, err := r.engineStatus(userID, backendURL)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	user := &backendv1.User{}
+	if err := r.Client.Get(ctx, req.NamespacedName, user); err != nil {
+		if apierrors.IsNotFound(err) {
+			// User is already deleted, kill engine if it's still running.
+			if userState == Running {
+				return ctrl.Result{}, r.killEngine(userID, backendURL)
+			}
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if !user.ObjectMeta.DeletionTimestamp.IsZero() {
+		// object is marked for deletion
+		if userState == Running {
+			return ctrl.Result{}, r.killEngine(userID, backendURL)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	user.Status.Backend = backendURL
+	defer r.Status().Update(ctx, user)
+	if userState == Running {
+		user.Status.State = string(Running)
+		return ctrl.Result{RequeueAfter: 20 * time.Second}, nil
+	}
+
+	if err = r.runEngine(user.Spec.ID, backendURL); err != nil {
+		user.Status.State = err.Error()
+		return ctrl.Result{RequeueAfter: 1 * time.Second}, nil
+	}
+
+	user.Status.State = string(Starting)
+	return ctrl.Result{RequeueAfter: 300 * time.Millisecond}, nil
+}
+
+func (r *UserReconciler) getBackendURL(name string, ctx context.Context, req ctrl.Request) (string, int64, error) {
+	uid, err := strconv.ParseInt(strings.TrimPrefix(name, "user-"), 10, 64)
+	if err != nil {
+		return "", 0, err
+	}
+	endpoints := &corev1.Endpoints{}
+	// fetch backend endpoints.
+	if err := r.Get(ctx, client.ObjectKey{Namespace: req.Namespace, Name: "backend"}, endpoints); err != nil {
+		return "", 0, err
+	}
+	// StatefulSet endpoints they are always ordered by name.
+	var backendURLs []string
+	for _, subset := range endpoints.Subsets {
+		for _, addr := range subset.Addresses {
+			urlStr := fmt.Sprintf("http://%s:%d", addr.IP, subset.Ports[0].Port)
+			backendURLs = append(backendURLs, urlStr)
+		}
+	}
+	return backendURLs[uid%int64(len(backendURLs))], uid, nil
+}
+
+func (r *UserReconciler) engineStatus(userID int64, userURL string) (UserState, error) {
+	resp, err := http.Get(fmt.Sprintf("%s/engineStatus?userid=%d", userURL, userID))
```

Comment:
Are these APIs fully synchronous? What's the concurrency mode here? Is the whole thing fully single-threaded? And are we worried that a stalled or broken worker can block the controller until timeout and prevent other users from running?

Reply:
My understanding is It synchronous per resource so it won't impact other users.

Reply:
Turns out the worker pool default size is 1.  
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controller.go#L38
I've updated the controller with 10 seconds timeout for the whole operation.